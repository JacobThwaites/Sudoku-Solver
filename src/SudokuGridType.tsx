export type SudokuGridType = Array<Array<number|null>>;