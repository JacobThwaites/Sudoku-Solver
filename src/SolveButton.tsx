import { Dispatch, SetStateAction } from "react";
import StyledButton from "./sudoku/StyledButton";
import { SudokuGridType } from "./SudokuGridType";
import { SudokuSolver } from "./solver/SudokuSolver";
import { isValidSudoku } from "./utils/utils";
import { copySudokuGrid } from "./utils/copySudokuGrid";

interface SolveButtonProps {
  disabled: boolean;
  sudoku: SudokuGridType;
  setStartingPosition: Function;
  updateRows: Function;
  setSolutionSteps: Function;
  setNoSolutionFound: Dispatch<SetStateAction<boolean>>;
  setIsInvalidPuzzle: Dispatch<SetStateAction<boolean>>;
}

export default function SolveButton(props: SolveButtonProps) {
  async function fetchSolution() {
    const startingPosition = copySudokuGrid(props.sudoku);
    props.setStartingPosition(startingPosition);

    if (!isValidSudoku(props.sudoku)) {
      props.setIsInvalidPuzzle(true);
      return;
    }

    const solver = new SudokuSolver(props.sudoku);
    const solution = solver.solve();

    if (!solution) {
      props.setNoSolutionFound(true);
    } else {
      props.updateRows(solution);
      props.setSolutionSteps(solver.steps);
    }
  }

  return (
    <StyledButton
      disabled={props.disabled}
      onClick={fetchSolution}
      label="Solve"
    />
  );
}