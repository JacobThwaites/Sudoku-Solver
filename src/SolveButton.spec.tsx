import renderer from 'react-test-renderer';
import SolveButton from './SolveButton';
import { SudokuGridType } from "./SudokuGridType";


it('renders SolveButton correctly', () => {
    const setStartingPosition = jest.fn();
    const sudoku: SudokuGridType = [];
    const updateRows = jest.fn();
    const setSolutionSteps = jest.fn();
    const setNoSolutionFound = jest.fn();
    const setIsInvalidPuzzle = jest.fn();

    const component = renderer.create(
        <SolveButton
            disabled={false}
            setStartingPosition={setStartingPosition}
            updateRows={updateRows}
            setSolutionSteps={setSolutionSteps}
            setNoSolutionFound={setNoSolutionFound}
            setIsInvalidPuzzle={setIsInvalidPuzzle}
            sudoku={sudoku}
        />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});