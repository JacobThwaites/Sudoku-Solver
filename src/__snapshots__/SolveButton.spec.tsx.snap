// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders SolveButton correctly 1`] = `
<button
  className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium styled-button css-sghohy-MuiButtonBase-root-MuiButton-root"
  disabled={false}
  onBlur={[Function]}
  onClick={[Function]}
  onContextMenu={[Function]}
  onDragLeave={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  tabIndex={0}
  type="button"
>
  Solve
</button>
`;
